internal/teststep/config.go

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package teststep
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	rawConfigFileName     = "terraform_plugin_test.tf"
+	rawConfigFileNameJSON = rawConfigFileName + ".json"
+)
+
+var (
+	providerConfigBlockRegex  = regexp.MustCompile(`provider "?[a-zA-Z0-9_-]+"? {`)
+	terraformConfigBlockRegex = regexp.MustCompile(`terraform {`)
+)
+
+type Config interface {
+	HasConfiguration() bool
+	HasConfigurationFiles() bool
+	HasProviderBlock(context.Context) (bool, error)
+	HasTerraformBlock(context.Context) (bool, error)
+	Write(context.Context, string) error
+}
+
+type configuration struct {
+	directory string
+	raw       string
+}
+
+type ConfigurationRequest struct {
+	Directory string
+	Raw       string
+}
+
+func Configuration(req ConfigurationRequest) (configuration, error) {
+	var populatedConfig []string
+	var config configuration
+
+	if req.Directory != "" {
+		populatedConfig = append(populatedConfig, fmt.Sprintf("%q", "directory"))
+
+		config = configuration{
+			directory: req.Directory,
+		}
+	}
+
+	if req.Raw != "" {
+		populatedConfig = append(populatedConfig, fmt.Sprintf("%q", "raw"))
+
+		config = configuration{
+			raw: req.Raw,
+		}
+	}
+
+	if len(populatedConfig) > 1 {
+		return configuration{}, fmt.Errorf(
+			"both %s are populated, only one configuration option is allowed",
+			strings.Join(populatedConfig, " and "),
+		)
+	}
+
+	return config, nil
+}
+
+func (c configuration) HasConfiguration() bool {
+	if c.directory != "" {
+		return true
+	}
+
+	if c.raw != "" {
+		return true
+	}
+
+	return false
+}
+
+func (c configuration) HasConfigurationFiles() bool {
+	return c.directory != ""
+}
+
+// HasProviderBlock returns true if the Config has declared a provider
+// configuration block, e.g. provider "examplecloud" {...}
+func (c configuration) HasProviderBlock(ctx context.Context) (bool, error) {
+	switch {
+	case c.hasRaw(ctx):
+		return providerConfigBlockRegex.MatchString(c.raw), nil
+	case c.hasDirectory(ctx):
```

Comment:
💡 One thing you could do here to avoid checking each case repetitively introduce two separate `configuration` implementations -- `ConfigurationDirectory` and `ConfigurationString`

Reply:
I've added `configurationDirectory` and `configurationString` types that implement the `Config` interface. The `Configuration()` constructor function now returns `Config` interface to accommodate the various types we will need (e.g., `configurationDirectory`, `configurationString`, `configurationFile`). Owing to the switch to returning an interface from the `Configuration()` constructor function, nil checks have been added.